src/dense_subgraph.rs

```diff
@@ -0,0 +1,215 @@
+    pub map: Vec<NodeIndex>,
+    pub subgraph: Vec<[NodeIndex; 2]>,
+}
+
+pub fn densest_subgraph<Ty>(
+    py: Python,
+    graph: &StablePyGraph<Ty>,
+    num_nodes: usize,
+    weight_callback: Option<PyObject>,
+) -> PyResult<(StablePyGraph<Ty>, NodeMap)>
+where
+    Ty: EdgeType + Sync,
+{
+    let node_indices: Vec<NodeIndex> = graph.node_indices().collect();
+    let float_callback =
+        |callback: PyObject, source_node: usize, target_node: usize| -> PyResult<f64> {
+            let res = callback.as_ref(py).call1((source_node, target_node))?;
+            res.extract()
+        };
+    let mut weight_map: Option<HashMap<[NodeIndex; 2], f64>> = None;
+
+    if weight_callback.is_some() {
+        let mut inner_weight_map: HashMap<[NodeIndex; 2], f64> =
+            HashMap::with_capacity(graph.edge_count());
+        let callback = weight_callback.as_ref().unwrap();
+        for edge in graph.edge_references() {
+            let source: NodeIndex = edge.source();
+            let target: NodeIndex = edge.target();
+            let weight = float_callback(callback.clone_ref(py), source.index(), target.index())?;
+            inner_weight_map.insert([source, target], weight);
+        }
+        weight_map = Some(inner_weight_map);
+    }
+    let reduce_identity_fn = || -> SubsetResult {
+        SubsetResult {
+            count: 0,
+            map: Vec::new(),
+            error: std::f64::INFINITY,
+            subgraph: Vec::new(),
+        }
+    };
+
+    let reduce_fn = |best: SubsetResult, curr: SubsetResult| -> SubsetResult {
+        if weight_callback.is_some() {
+            if curr.count >= best.count && curr.error <= best.error {
+                curr
+            } else {
+                best
+            }
+        } else if curr.count > best.count {
+            curr
+        } else {
+            best
+        }
+    };
+
+    let best_result = node_indices
+        .into_par_iter()
+        .map(|index| {
+            let mut subgraph: Vec<[NodeIndex; 2]> = Vec::with_capacity(num_nodes);
+            let mut bfs = Bfs::new(&graph, index);
+            let mut bfs_vec: Vec<NodeIndex> = Vec::with_capacity(num_nodes);
+            let mut bfs_set: HashSet<NodeIndex> = HashSet::with_capacity(num_nodes);
+
+            let mut count = 0;
+            while let Some(node) = bfs.next(&graph) {
+                bfs_vec.push(node);
+                bfs_set.insert(node);
+                count += 1;
+                if count >= num_nodes {
+                    break;
+                }
+            }
+            let mut connection_count = 0;
+            for node in &bfs_vec {
+                for j in graph.node_indices().filter(|j| bfs_set.contains(j)) {
+                    if graph.contains_edge(*node, j) {
+                        connection_count += 1;
+                        subgraph.push([*node, j]);
+                    }
+                }
+            }
+            let error = match &weight_map {
```

Comment:
TBH, I don't remember if I did it on purpose or not. I feel like I might have because it would require an additional callback. But I'll add another callback function for this now

Reply:
Done in: https://github.com/Qiskit/rustworkx/pull/635/commits/40aac33693308115a4a88ee714dffd8830398880